hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -99,63 +100,22 @@ sealed abstract class Block extends Product with AutoLocated:
   def floatOutDefns =
-    def rec(b: Block, acc: List[Defn]): (Block, List[Defn]) = b match
-      case Match(scrut, arms, dflt, rest) =>
-        val (armsRes, armsDefns) = arms.foldLeft[(List[(Case, Block)], List[Defn])](Nil, acc)(
-          (accc, d) =>
-            val (accCases, accDefns) = accc
-            val (cse, blk) = d
-            val (resBlk, resDefns) = rec(blk, accDefns)
-            ((cse, resBlk) :: accCases, resDefns)
-        )
-        dflt match
-        case None =>
-          val (rstRes, rstDefns) = rec(rest, armsDefns)
-          (Match(scrut, armsRes, None, rstRes), rstDefns)
-
-        case Some(dflt) =>
-          val (dfltRes, dfltDefns) = rec(dflt, armsDefns)
-          val (rstRes, rstDefns) = rec(rest, dfltDefns)
-          (Match(scrut, armsRes, S(dfltRes), rstRes), rstDefns)
-        
-      case Return(res, implct) => (b, acc)
-      case Throw(exc) => (b, acc)
-      case Label(label, body, rest) =>
-        val (bodyRes, bodyDefns) = rec(body, acc)
-        val (rstRes, rstDefns) = rec(rest, bodyDefns)
-        (Label(label, bodyRes, rstRes), rstDefns)
-      case Break(label) => (b, acc)
-      case Continue(label) => (b, acc)
-      case Begin(sub, rest) => 
-        val (subRes, subDefns) = rec(sub, acc)
-        val (rstRes, rstDefns) = rec(rest, subDefns)
-        (Begin(subRes, rstRes), rstDefns)
-      case TryBlock(sub, finallyDo, rest) =>
-        val (subRes, subDefns) = rec(sub, acc)
-        val (finallyRes, finallyDefns) = rec(rest, subDefns)
-        val (rstRes, rstDefns) = rec(rest, finallyDefns)
-        (TryBlock(subRes, finallyRes, rstRes), rstDefns)
-      case Assign(lhs, rhs, rest) => 
-        val (rstRes, rstDefns) = rec(rest, acc)
-        (Assign(lhs, rhs, rstRes), rstDefns)
-      case a @ AssignField(path, nme, result, rest) =>
-        val (rstRes, rstDefns) = rec(rest, acc)
-        (AssignField(path, nme, result, rstRes)(a.symbol), rstDefns)
-      case Define(defn, rest) => defn match
-        case ValDefn(owner, k, sym, rhs) => 
-          val (rstRes, rstDefns) = rec(rest, acc)
-          (Define(defn, rstRes), rstDefns)
-        case _ =>
-          val (rstRes, rstDefns) = rec(rest, defn :: acc)
-          (rstRes, rstDefns)
-      case HandleBlock(lhs, res, par, cls, handlers, body, rest) =>
-        val (rstRes, rstDefns) = rec(rest, acc)
-        (HandleBlock(lhs, res, par, cls, handlers, body, rstRes), rstDefns)
-      case HandleBlockReturn(res) => (b, acc)
-      case End(msg) => (b, acc)
-    rec(this, Nil)
-  
+    val defns = ListBuffer[Defn]()
+    val transformer = new BlockTransformerShallow(SymbolSubst()):
+      override def applyBlock(b: Block): Block = b match
+        case Define(defn, rest) => defn match
+          case v: ValDefn => super.applyBlock(b)
+          case _ =>
+            defns.addOne(defn)
+            applyBlock(rest)
+        case _ => super.applyBlock(b)
+    
+    (transformer.applyBlock(this), defns.reverse.toList)
```

Comment:
If you're going to reverse it at the end, there's no need to use `ListBuffer`. Just build the reversed list directly!
```suggestion
    var defns: List[Defn] = Nil
    val transformer = new BlockTransformerShallow(SymbolSubst()):
      override def applyBlock(b: Block): Block = b match
        case Define(defn, rest) => defn match
          case v: ValDefn => super.applyBlock(b)
          case _ =>
            defns ::= defn
            applyBlock(rest)
        case _ => super.applyBlock(b)
    
    (transformer.applyBlock(this), defns.toList)
```